### PR TITLE
Extend AMQP support by supporting delete and create event

### PR DESCRIPTION
**Pull request by coolo**
Also fetch jobs that were updated in between watcher restarts

### CODE REVIEW

File: lib/Dashboard/Model/AMQP.pm [hunk 1]
```diff
@@ -18,7 +18,17 @@ use Mojo::Base -base, -signatures;
 
 has [qw(log jobs)];
 
-sub handle ($self, $key, $data) {
+# called from startup in AMQP watcher to catch up with jobs started/ended while the watcher wasn't alive
+sub fetch_old ($self) {
+  my $db = $self->pg->db;
+  my $id = $db->query("SELECT min(job_id) from openqa_jobs where status='waiting'")->hash->{min};
```
Comment (kraih) [inline lib/Dashboard/Model/AMQP.pm:24]: That doesn't work, if there's no result you get `undef` back and can't access the hash for `->{min}`.

File: lib/Dashboard/Model/Jobs.pm [hunk 2]
```diff
@@ -31,6 +31,26 @@ sub add ($self, $job) {
   );
 }
 
+sub delete_job ($self, $id) {
+  $self->pg->db('DELETE from openqa_jobs where job_id=?', $id);
+}
+
+# called from startup in AMQP watcher to catch up with jobs started/ended while the watcher wasn't alive
+sub fetch_old ($self) {
```
Comment (kraih) [inline lib/Dashboard/Model/Jobs.pm:39]: This is duplicate code from `lib/Dashboard/Model/AMQP.pm`, right?

File: lib/Dashboard/Model/Jobs.pm [hunk 3]
```diff
@@ -31,6 +31,26 @@ sub add ($self, $job) {
...
+  my $db = $self->pg->db;
+  my $id = $db->query("SELECT min(job_id) from openqa_jobs where status='waiting'")->hash->{min};
```
Comment (kraih) [inline lib/Dashboard/Model/Jobs.pm:41]: Same issue as before, return value would be `undef` without a result.

File: lib/Dashboard/Model/OpenQA.pm [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+}
+
+# for easier mocking - we may want to make this configurable, but 50 jobs seem to be a good compromise
+# and the value isn't free to choose, around 80 jobs will already create too long URLs
+sub _delta ($self) {
+  return 50;
+}
+
+sub _openqa_url ($self) {
+  return Mojo::URL->new('https://openqa.suse.de');
```
Comment (kraih) [inline lib/Dashboard/Model/OpenQA.pm:76]: This should use the URL from the app config.

File: lib/Dashboard/Model/OpenQA.pm [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+  my $delta = $self->_delta;
+  $self->log->info("Fetching $delta jobs starting $id");
+
+  my $url = $self->_openqa_url->path('/api/v1/jobs')->query(ids => [($id .. $id + $delta)]);
+  $self->ua->get_p($url => {Accept => 'application/json'})->then(
+    sub ($tx) {
+
+      my %seen_jobs;
+      if ($tx->error) {
+        $self->log->error(Mojo::Util::dumper($tx));
```
Comment (kraih) [inline lib/Dashboard/Model/OpenQA.pm:45]: We probably shouldn't dump the whole transaction in production.

File: lib/Dashboard/Model/AMQP.pm (hunk 1 above)
Comment (coolo) [reply lib/Dashboard/Model/AMQP.pm:24]: This is brute force anyway and I guess we should be smarter about this. I just don't know how :)

In production DB we have a waiting job with an id 2M away from HEAD - so we would fetch 2M jobs to fix up.

File: lib/Dashboard/Model/Jobs.pm (hunk 2 above)
Comment (coolo) [reply lib/Dashboard/Model/Jobs.pm:39]: I moved the code forth and back as I couldn't find a sweet spot between the models dependencies. The code in AMQP is dead code, I'll remove that

**Comment by kalikiana** [comment]
@coolo I think you need to rebase. Not sure I can offer much to help, just noticed this one browsing pending PRs

**Pull request closed without merging**